File: LeetCode/src/stack/RemovingStarsFromString.java
package stack;

import java.util.Stack;

public class RemovingStarsFromString {
    public String removeStars(String s) {
        Stack<Character> stack = new Stack<>();

        for (char ch : s.toCharArray()) {
            if (ch == '*') {
                if (stack.isEmpty()) {
                    continue;
                }
                stack.pop();
            } else {
                stack.push(ch);
            }
        }

        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.append(stack.pop());
        }

        return sb.reverse().toString();
    }

    public static void main(String[] args) {
        RemovingStarsFromString solution = new RemovingStarsFromString();
        String s1 = "leet**cod*e";
        System.out.println(solution.removeStars(s1)); // Output: lecoe

        String s2 = "erase*****";
        System.out.println(solution.removeStars(s2)); // Output: ""
    }
}

File: LeetCode_Python/stack/RemovingStarsFromString.py
class RemovingStarsFromString:
    def removeStars(self, s: str) -> str:
        stack = []

        for char in s:
            if char == "*":
                if stack:
                    stack.pop()
            else:
                stack.append(char)

        return "".join(stack)


# Time Complexity: O(n), where n is the length of the input string
# Space Complexity: O(n), where n is the length of the input string

if __name__ == "__main__":
    solution = RemovingStarsFromString()

    print("Expected Output: lecoe")
    print("Actual Output:", solution.removeStars("leet**cod*e"))

    print("Expected Output: ")
    print("Actual Output:", solution.removeStars("erase*****"))
